src/base/bittorrent/session.cpp

```diff
@@ -1163,80 +1163,51 @@ void Session::applyBandwidthLimits(lt::settings_pack &settingsPack) const
-    Q_ASSERT(m_metricIndices.net.recvIPOverheadBytes >= 0);
-
-    m_metricIndices.net.sentTrackerBytes = lt::find_metric_idx("net.sent_tracker_bytes");
-    Q_ASSERT(m_metricIndices.net.sentTrackerBytes >= 0);
-
-    m_metricIndices.net.recvTrackerBytes = lt::find_metric_idx("net.recv_tracker_bytes");
-    Q_ASSERT(m_metricIndices.net.recvTrackerBytes >= 0);
-
-    m_metricIndices.net.recvRedundantBytes = lt::find_metric_idx("net.recv_redundant_bytes");
-    Q_ASSERT(m_metricIndices.net.recvRedundantBytes >= 0);
-
-    m_metricIndices.net.recvFailedBytes = lt::find_metric_idx("net.recv_failed_bytes");
-    Q_ASSERT(m_metricIndices.net.recvFailedBytes >= 0);
-
-    m_metricIndices.peer.numPeersConnected = lt::find_metric_idx("peer.num_peers_connected");
-    Q_ASSERT(m_metricIndices.peer.numPeersConnected >= 0);
-
-    m_metricIndices.peer.numPeersDownDisk = lt::find_metric_idx("peer.num_peers_down_disk");
-    Q_ASSERT(m_metricIndices.peer.numPeersDownDisk >= 0);
-
-    m_metricIndices.peer.numPeersUpDisk = lt::find_metric_idx("peer.num_peers_up_disk");
-    Q_ASSERT(m_metricIndices.peer.numPeersUpDisk >= 0);
-
-    m_metricIndices.dht.dhtBytesIn = lt::find_metric_idx("dht.dht_bytes_in");
-    Q_ASSERT(m_metricIndices.dht.dhtBytesIn >= 0);
-
-    m_metricIndices.dht.dhtBytesOut = lt::find_metric_idx("dht.dht_bytes_out");
-    Q_ASSERT(m_metricIndices.dht.dhtBytesOut >= 0);
-
-    m_metricIndices.dht.dhtNodes = lt::find_metric_idx("dht.dht_nodes");
-    Q_ASSERT(m_metricIndices.dht.dhtNodes >= 0);
-
-    m_metricIndices.disk.diskBlocksInUse = lt::find_metric_idx("disk.disk_blocks_in_use");
-    Q_ASSERT(m_metricIndices.disk.diskBlocksInUse >= 0);
-
-    m_metricIndices.disk.numBlocksRead = lt::find_metric_idx("disk.num_blocks_read");
-    Q_ASSERT(m_metricIndices.disk.numBlocksRead >= 0);
-
-    m_metricIndices.disk.numBlocksCacheHits = lt::find_metric_idx("disk.num_blocks_cache_hits");
-    Q_ASSERT(m_metricIndices.disk.numBlocksCacheHits >= 0);
-
-    m_metricIndices.disk.writeJobs = lt::find_metric_idx("disk.num_write_ops");
-    Q_ASSERT(m_metricIndices.disk.writeJobs >= 0);
-
-    m_metricIndices.disk.readJobs = lt::find_metric_idx("disk.num_read_ops");
-    Q_ASSERT(m_metricIndices.disk.readJobs >= 0);
-
-    m_metricIndices.disk.hashJobs = lt::find_metric_idx("disk.num_blocks_hashed");
-    Q_ASSERT(m_metricIndices.disk.hashJobs >= 0);
-
-    m_metricIndices.disk.queuedDiskJobs = lt::find_metric_idx("disk.queued_disk_jobs");
-    Q_ASSERT(m_metricIndices.disk.queuedDiskJobs >= 0);
+    const auto findMetricIdx = [](const char *name) -> int
+    {
+        const int index = lt::find_metric_idx(name);
+        Q_ASSERT(index >= 0);
+        return index;
+    };
 
-    m_metricIndices.disk.diskJobTime = lt::find_metric_idx("disk.disk_job_time");
-    Q_ASSERT(m_metricIndices.disk.diskJobTime >= 0);
+    m_metricIndices =
+    {
+        {
+            findMetricIdx("net.has_incoming_connections"),
```

Comment:
I dislike this approach. Previous one clearly shows that you assign indexes to appropriate fields. Now it is very easy to confuse something. Although this code rarely changes, it still creates a sense of insecurity...

Reply:
>Now it is very easy to confuse something. 

Like fields mismatch? it might happen when not careful. But in the old "member assignment" way, it is still possible to miss some fields when `struct` layout changes.

Actually I like this way, it is very structured, it shows we are creating a temporary `struct` and then use it (via `operator=`) to replace the old `struct` systematically.

Reply:
>But in the old "member assignment" way, it is still possible to miss some when adding additional fields to `struct`.

But isn't it possible by using initializer list as well?

Reply:
The problem of this change is that we no longer have visible mapping between metric names and appropriate variable names.

Reply:
>But isn't it possible by using initializer list as well?

Exactly, so the 2 styles aren't that much different on this mismatch problem.

>The problem of this change is that we no longer have visible mapping between metric names and appropriate variable names.

I don't know why the "explicit mapping" is required. It is quite often (in other programming languages) that pattern matching is being used like our case here. The upside: more terse code.
Somehow related, C++20 added "designated initializers" which shows the trend here.

As for verifying the fields are correctly filled, I still need to open 2 windows, one for the `struct` definition and the other for filling the values. This is the same for 2 styles, at least for me.

I'm fine with reverting it back if you insist, but before that I want to know if this style is ok for you:
```c++
m_metricIndices.disk =
{
    // ...
};
m_metricIndices.dht=
{
    // ...
};
```

Reply:
>I don't know why the "explicit mapping" is required.

Just to see explicitly that, for example, the variable `m_metricIndices.net.hasIncomingConnections` is assigned with an index for the metric "net.has_incoming_connections". Doesn't this make it easier to maintain the code?
>C++20 added "designated initializers" which shows the trend here (and probably addresses your concern).

I agree. In C++20 it would have sufficient readability, as before. But until then, I wouldn't change it to aggregate initialization.

>I'm fine with reverting it back if you insist

Okay, I don't insist... I just feel doubtful about it. I'll leave it to you.
>before that I want to know if this style is ok for you

No. Not these pointless half-measures. Either a completely old approach, or a completely new one.